fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java

```diff
@@ -0,0 +1,289 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation.materialization;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.starrocks.analysis.Expr;
+import com.starrocks.catalog.Function;
+import com.starrocks.catalog.FunctionSet;
+import com.starrocks.catalog.ScalarType;
+import com.starrocks.catalog.Type;
+import com.starrocks.common.Pair;
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CastOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ScalarOperatorRewriter;
+import com.starrocks.sql.optimizer.rewrite.scalar.ImplicitCastRule;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.starrocks.catalog.Function.CompareMode.IS_IDENTICAL;
+
+/**
+ * `AggregateFunctionRewriter` will try to rewrite some agg functions to some transformations so can be
+ * better to be rewritten.
+ * eg: AVG -> SUM / COUNT
+ * eg: COUNT(DISTINCT) -> BITMAP_COUNT(BITMAP_UNION(TO_BITMAP)))
+ */
+public class AggregateFunctionRewriter {
+    final ScalarOperatorRewriter scalarRewriter = new ScalarOperatorRewriter();
+    final ColumnRefFactory queryColumnRefFactory;
+
+    // new rewrite agg can reuse old agg functions if it has existed.
+    final Map<ColumnRefOperator, CallOperator> oldAggregations;
+
+    Map<ColumnRefOperator, CallOperator> newColumnRefToAggFuncMap;
+
+    public AggregateFunctionRewriter(ColumnRefFactory queryColumnRefFactory,
+                                     Map<ColumnRefOperator, CallOperator> oldAggregations) {
+        this.queryColumnRefFactory = queryColumnRefFactory;
+        this.oldAggregations = oldAggregations;
+    }
+
+    public AggregateFunctionRewriter(ColumnRefFactory queryColumnRefFactory,
+                                     Map<ColumnRefOperator, CallOperator> oldAggregations,
+                                     Map<ColumnRefOperator, CallOperator> newColumnRefToAggFuncMap) {
+        this.queryColumnRefFactory = queryColumnRefFactory;
+        this.oldAggregations = oldAggregations;
+        this.newColumnRefToAggFuncMap = newColumnRefToAggFuncMap;
+    }
+
+    public boolean canRewriteAggFunction(ScalarOperator op) {
+        if (!(op instanceof CallOperator)) {
+            return false;
+        }
+
+        CallOperator aggFunc = (CallOperator) op;
+        String aggFuncName = aggFunc.getFnName();
+        if (aggFuncName.equals(FunctionSet.AVG)) {
+            return true;
+        }
+        // BITMAP
+        if (aggFuncName.equals(FunctionSet.COUNT) && aggFunc.isDistinct()) {
```

Comment:
use a set, put all supported funtion names into it, check with contain?

Reply:
Let me think it.